scripts/west_commands/patch.py

```diff
@@ -0,0 +1,332 @@
+            expect_sha256 = patch_info["sha256sum"]
+            hasher = hashlib.sha256()
+            hasher.update(patch_file_data)
+            actual_sha256 = hasher.hexdigest()
+            if actual_sha256 != expect_sha256:
+                self.dbg("FAIL")
+                self.err(
+                    f"sha256 mismatch for {pth}:\n"
+                    f"expect: {expect_sha256}\n"
+                    f"actual: {actual_sha256}"
+                )
+                failed_patch = pth
+                break
+            self.dbg("OK")
+            patch_count += 1
+            patch_file_data = None
+
+            mod = patch_info["module"]
+            mod_path = Path(args.west_workspace) / mod
+            patched_mods.add(mod)
+
+            self.dbg(f"patching {mod}... ", end="")
+            origdir = os.getcwd()
+            os.chdir(mod_path)
+            apply_cmd += patch_path
+            apply_cmd_list.extend([patch_path])
+            proc = subprocess.run(apply_cmd_list)
+            if proc.returncode:
+                self.dbg("FAIL")
+                self.err(proc.stderr)
+                failed_patch = pth
+            self.dbg("OK")
+            os.chdir(origdir)
+
+        if not failed_patch:
+            self.inf(f"{patch_count} patches applied successfully \\o/")
+            return
+
+        if args.roll_back:
+            self.clean(args, yml, patched_mods)
+
+        self.die(f"failed to apply patch {pth}")
+
+    def clean(self, args, yml, mods=None):
+        git_clean_cmd = yml["clean-command"]
+        git_checkout_cmd = yml["checkout-command"]
+
+        git_clean_cmd_list = shlex.split(git_clean_cmd)
+        git_checkout_cmd_list = shlex.split(git_checkout_cmd)
+
+        if not git_clean_cmd_list and not git_checkout_cmd_list:
+            self.dbg("no clean or checkout commands specified")
+            return
+
+        origdir = os.getcwd()
+        for mod, mod_path in Patch.get_mod_paths(args, yml).items():
+            if mods and mod not in mods:
+                continue
+            try:
+                self.dbg(f"Cleaning {mod}")
+                os.chdir(mod_path)
+                proc = subprocess.run(
+                    git_checkout_cmd_list, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
+                )
+                if proc.returncode:
+                    self.err(f"{git_checkout_cmd_list} failed for {mod}: {proc.returncode}")
+                proc = subprocess.run(
+                    git_clean_cmd_list, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
+                )
+                if proc.returncode:
+                    self.err(f"{git_clean_cmd_list} failed for {mod}: {proc.returncode}")
+            except Exception as e:
+                # If this fails for some reason, just log it and continue
+                self.err(f"failed to clean up {mod}: {e}")
+
+        os.chdir(origdir)
+
+    def list(self, args, yml):
+        for patch_info in yml["patches"]:
```

Comment:
Doesn't this throw an error if there's no entry?

```suggestion
        for patch_info in yml.get("patches", []):
```

Reply:
I assumed that by not specifying `allowempty`, that patches must be populated, otherwise YAML validation should fail.

I can test that theory.

Indeed, an empty patches object (`patches: {}`) will not work / is considered broken YAML.
```shell
west patch list        
FATAL ERROR: ERROR: Malformed yaml /Users/cfriedt/zephyrproject/XXXX/zephyr/patches.yml: <SchemaError: error code 2: Schema validation failed:
 - Value '{}' is not a list. Value path: '/patches'.: Path: '/'>
```

Reply:
Oh wait - I added a couple of extra checks.

One of them looks for a completely empty file (that triggers an exception in pykwalify), and then it looks like my test above was not correct, so I'll use your suggestion.

Reply:
Fixed!